src/checkers/inference/model/ComparableConstraint.java

```diff
@@ -11,18 +13,38 @@
     private final Slot first;
     private final Slot second;
 
-    protected ComparableConstraint(Slot first, Slot second, AnnotationLocation location) {
+    private ComparableConstraint(Slot first, Slot second, AnnotationLocation location) {
         super(Arrays.asList(first, second), location);
         this.first = first;
         this.second = second;
     }
 
-    protected ComparableConstraint(Slot first, Slot second) {
+    private ComparableConstraint(Slot first, Slot second) {
         super(Arrays.asList(first, second));
         this.first = first;
         this.second = second;
     }
 
+    protected static Constraint create(Slot first, Slot second, AnnotationLocation location,
+            QualifierHierarchy realQualHierarchy) {
+        if (first == null || second == null) {
+            ErrorReporter.errorAbort("Create comparable constraint with null argument. Subtype: "
+                    + first + " Supertype: " + second);
+        }
+
+        if (first instanceof ConstantSlot && second instanceof ConstantSlot) {
```

Comment:
Isn't the logic of this if block the job of `ConstraintManager`? In order to create a `ComparableConstraint`, why do I need to care if they are two `ConstantSlot` case in `ComparableConstraint` class? Why should a "data model"(`ComparableConstraint`) care about business logic(if two slots are two Constants)? Seems decreasing cohesion in `ConstraintManager`

Reply:
Instead of having constraint-constraint checks everywhere as the other PRs would have result in, for subtype constraint especially, the point is to centralize and make the constraints themselves be responsible for normalization.

The manager delegates, and is still responsible for collecting the set of constraints and issuing errors.